### PR TITLE
[TTS][SpeechLLM] Merge ASR SFT changes to SpeechLLM T5

### PR DESCRIPTION
**Pull request by subhankar-ghosh**
# What does this PR do ?

ASR SFT related code changes.

**Collection**: [Note which collection this PR will affect]
NLM

# Changelog 
At some places the flow needs to go into an if-else statement based on the task at hand - ASR or TTS.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/conf/megatron_t5_speechlm_infer.yaml
Comment (shehzeen) [inline examples/nlp/language_modeling/conf/megatron_t5_speechlm_infer.yaml:1]: megatron_t5_speechlm_inference.yaml already exists. We should just keep one of them since they are almost the same.

File: examples/nlp/language_modeling/megatron_t5_speechlm_infer.py
Comment (shehzeen) [inline examples/nlp/language_modeling/megatron_t5_speechlm_infer.py:1]: Same here, megatron_t5_speechlm_sft_inference.py already exists. We can update megatron_t5_speechlm_sft_inference.py to be the same as this and keep just one of them.

File: nemo/collections/nlp/data/language_modeling/megatron/t5_speechlm_dataset.py [hunk 1]
```diff
@@ -191,11 +192,15 @@ def load_data(self, dataset):
                 if self.train_task != 'tts':
```
Comment (shehzeen) [inline nemo/collections/nlp/data/language_modeling/megatron/t5_speechlm_dataset.py:192]: If train_task is None, does that mean it will train on all tasks in the manifest? Or what should train_task be if we want to train it on both ASR and TTS?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py [hunk 2]
```diff
@@ -360,50 +360,70 @@ def fwd_output_and_loss_func(dataloader_iter, model):
...
-                        all_layer_tokens = torch.stack([first_layer_tokens] + other_layer_tokens)  # (8, t)
-                        all_layer_tokens = self.convert_tokens_to_range(
-                            all_layer_tokens, apply_offset_correction=False
-                        )
-                        all_layer_tokens = torch.clip(all_layer_tokens, 0, 1023)
-                        predicted_wav = self.additional_models['encodec'].decode([[all_layer_tokens[None], None]])[
-                            0, 0
-                        ]
-                        self.logger.experiment.add_audio("Pred Wav", predicted_wav, self.global_step, 24000)
+                        if torch.count_nonzero(speech_mask) == 0:
```
Comment (shehzeen) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:363]: When we train on both ASR and TTS, some rows in speech_mask will be zeros and some will be ones. We should consider slicing the batch into text outputs and speech outputs (as per speech_mask) and then log wer or audio. Another option is to compute WER only for batch indices where speech mask is zero and log audio for any index where speech_mask is one.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py [hunk 3]
```diff
@@ -576,25 +597,29 @@ def validation_step(self, batch, batch_idx, inference=False):
...
-            correct_predictions_i = speech_preds_i == labels_i  # (bs,t)
-            correct_predictions_i = correct_predictions_i * loss_mask * speech_mask  # (bs,t)
-            total_correct_predictions_i = torch.sum(correct_predictions_i)
-            total_predictions_i = torch.sum(loss_mask * speech_mask)
-            speech_accuracy_i = total_correct_predictions_i / total_predictions_i
-            loss_i = torch.nn.functional.cross_entropy(
-                speech_logits_i.permute(1, 2, 0), labels_i, reduction='none'
-            )  # (bs,t)
-            loss_i = torch.sum(loss_i * loss_mask * speech_mask) / total_predictions_i
+            if torch.count_nonzero(speech_mask) > 0:
```
Comment (shehzeen) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:606]: Same comment here. Let's try to log validation metrics for both ASR and TTS using speech_mask.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 4]
```diff
@@ -652,7 +652,11 @@ def forward(
 
                     # speech_logits_list will be used in loss calculation (parallel output)
                     speech_logits_list = []
-                    if self.seq_pattern in ["parallel", "delay_parallel"]:
+                    speech_logits = torch.zeros(
+                        [*token_logits.shape[:-1], self.speech_codebook_size, speech_layers],
+                        device=token_logits.device,
+                    )
+                    if self.seq_pattern in ["parallel", "delay_parallel"] and torch.count_nonzero(speech_mask) > 0:
```
Comment (shehzeen) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:659]: same issue here, if we train jointly on ASR and TTS task.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 5]
```diff
@@ -652,7 +652,11 @@ def forward(
 
                     # speech_logits_list will be used in loss calculation (parallel output)
                     speech_logits_list = []
-                    if self.seq_pattern in ["parallel", "delay_parallel"]:
+                    speech_logits = torch.zeros(
```
Comment (paarthneekhara) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:655]: speech_logits should be an unused variable now, so we can delete this I think.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py (hunk 2 above)
Comment (subhankar-ghosh) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:363]: This PR is just to add ASR as another SFT task and not a joint task. Currently, we need train_task to be either 'tts' or 'asr'. We cannot do joint training (asr and tts in the same batch) right now, what we can do is train for tts in one run and use the saved checkpoint to do ASR and so on. Having joint training (asr and tts in the same batch) might be trickier and might need changes to the dataloader as well.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py (hunk 3 above)
Comment (subhankar-ghosh) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:606]: Same reply as above.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 4 above)
Comment (subhankar-ghosh) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:659]: same.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 5 above)
Comment (subhankar-ghosh) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:655]: sure.
